src/resolver.js

```diff
@@ -65,56 +75,90 @@ const multihash = promisify((ipfs, path, callback) => {
+        let cidOfNextFile
         const nextFileName = item
 
-        for (let link of dagNode.links) {
-          if (link.name === nextFileName) {
-            // found multihash of requested named-file
-            multihashOfNextFile = mh.toB58String(link.multihash)
-            log('found multihash: ', multihashOfNextFile)
-            break
+        try {
+          for (let link of dagNode.links) {
+            if (link.name === nextFileName) {
+              // found multihash/cid of requested named-file
+              try {
+                // assume a Buffer with a valid CID
+                // (cid is allowed instead of multihash since https://github.com/ipld/js-ipld-dag-pb/pull/80)
+                cidOfNextFile = new CID(link.multihash)
+              } catch (err) {
+                // fallback to multihash
+                cidOfNextFile = new CID(mh.toB58String(link.multihash))
+              }
+              break
+            }
           }
+        } catch (err) {
+          return next(err)
         }
 
-        if (!multihashOfNextFile) {
-          return next(new Error(`no link named "${nextFileName}" under ${memo}`))
+        if (!cidOfNextFile) {
+          const missingLinkErr = new Error(`no link named "${nextFileName}" under ${memo}`)
+          missingLinkErr.parentDagNode = memo
+          missingLinkErr.missingLinkName = nextFileName
+          return next(missingLinkErr)
         }
 
-        next(null, multihashOfNextFile)
+        next(null, cidOfNextFile)
       })
-    }, (err, result) => {
+    }, (err, cid) => {
       if (err) {
         return callback(err)
       }
 
-      let cid
       try {
-        cid = new CID(mh.fromB58String(result))
+        cid = new CID(cid)
       } catch (err) {
         return callback(err)
       }
 
+      if (cid.codec === 'raw') {
+        // no need for additional lookup, its raw data
+        callback(null, { cid })
+      }
+
       ipfs.dag.get(cid, (err, dagResult) => {
         if (err) {
           return callback(err)
         }
 
-        let dagDataObj = Unixfs.unmarshal(dagResult.value.data)
-        if (dagDataObj.type === 'directory') {
-          let isDirErr = new Error('This dag node is a directory')
-          // add memo (last multihash) as a fileName so it can be used by directory
-          isDirErr.fileName = result
-          return callback(isDirErr)
+        try {
+          let dagDataObj = Unixfs.unmarshal(dagResult.value.data)
+          // There are at least two types of directories:
+          // - "directory"
+          // - "hamt-sharded-directory" (example: QmT5NvUtoM5nWFfrQdVrFtvGfKFmG7AHE8P34isapyhCxX)
+          if (dagDataObj.type === 'directory' || dagDataObj.type === 'hamt-sharded-directory') {
+            let isDirErr = new Error('This dag node is a directory')
+            // store memo of last multihash so it can be used by directory
+            isDirErr.cid = isDirErr.fileName = cid
```

Comment:
Backward compatibility: [js-ipfs uses `fileName` field](https://github.com/ipfs/js-ipfs/blob/4df45d1830e5e67915db941daa7ce906063a3647/src/http/gateway/resources/gateway.js#L56), so we are  keeping it here for now. 
When `ipfs-http-response` is released with these changes, will PR js-ipfs to switch from `fileName` to `cid`